### PR TITLE
Implement first version of beamchmark

### PR DESCRIPTION
**Pull request by mickel8**
related to membraneframework/membrane_rtc_engine#35

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,11 @@
+root = true
```
Comment (Qizot) [inline .editorconfig]: isn't it your personal file?

File: README.md [hunk 2]
```diff
@@ -1,6 +1,9 @@
 # Beamchmark
+[![Hex.pm](https://img.shields.io/hexpm/v/beamchmark.svg)](https://hex.pm/packages/beamchmark)
+[![API Docs](https://img.shields.io/badge/api-docs-yellow.svg?style=flat)](https://hexdocs.pm/beamchmark)
+[![CircleCI](https://circleci.com/gh/membraneframework/beamchmark.svg?style=svg)](https://circleci.com/gh/membraneframework/beamchmark)
 
-**TODO: Add description**
+Tool for measuring EVM performance.
```
Comment (Qizot) [inline README.md:6]: I would add some text description what gets measured and how in the readme, it is not that easy to read the numbers below.

File: lib/beamchmark.ex [hunk 3]
```diff
@@ -1,18 +1,198 @@
...
+  @doc """
+  Runs scenario and benchmarks EVM performance. 
 
+  Subsequent invocation of this function will also compare results with the previous ones.
   """
-  def hello do
-    :world
+  @spec run(Beamchmark.Scenario, options_t()) :: :ok
+  def run(scenario, opts) do
+    output_dir = opts[:output_dir] || @default_output_dir
```
Comment (Qizot) [inline lib/beamchmark.ex:63]: nitpick: `Keyword.get(opts, :output_dir, @default_output_dir)` looks nicer (at least for me).

File: lib/beamchmark.ex [hunk 4]
```diff
@@ -1,18 +1,198 @@
...
+    :scheduler.utilization(opts[:duration] || 60)
+    |> Beamchmark.SchedulerInfo.from_sched_util_result()
+  end
+
+  defp save(results, new_dir) do
+    File.mkdir_p!(new_dir)
+    out = Path.join(new_dir, @results_file_name)
+    results = :erlang.term_to_binary(results)
+    File.write!(out, results)
+    :noop
```
Comment (Qizot) [inline lib/beamchmark.ex]: why do you return `:noop`? I would guess that it means `no operation` and in this case you did some work

File: lib/beamchmark.ex [hunk 5]
```diff
@@ -1,18 +1,198 @@
...
+  end
+
+  defp print(new, base_dir) do
+    print_system_info()
+
+    base =
+      if File.exists?(base_dir) do
+        File.read!(Path.join(base_dir, @results_file_name))
```
Comment (Qizot) [inline lib/beamchmark.ex]: this could be futher divided to use pipe

File: lib/beamchmark/scheduler_info.ex [hunk 6]
```diff
@@ -0,0 +1,133 @@
+defmodule Beamchmark.SchedulerInfo do
+  @moduledoc false
+  # module representing different statistics about scheduler usage 
+
+  use Bunch.Access
+
+  alias __MODULE__
```
Comment (Qizot) [inline lib/beamchmark/scheduler_info.ex]: That alias does noting

**Review by Qizot — CHANGES_REQUESTED** [review]
I think that we are lacking the description of what the benchmark measures and how, why we need that package and how is it better than other solutions. I don't think that the example in readme is just enough as it requires all the people to decode it to be honest.  Not everyone has to know that there are different kinds of schedulers etc.

File: lib/beamchmark/scheduler_info.ex (hunk 6 above)
Comment (mickel8) [reply lib/beamchmark/scheduler_info.ex]: Actually it's needed for typespecs and documentation to make dialyzer happy

File: .editorconfig (hunk 1 above)
Comment (mickel8) [reply .editorconfig]: It's taken from `membrane_template_plugin`

Comment (Qizot) [reply .editorconfig]: Good to know that we have to install a plugin to make use of it :D

File: lib/beamchmark.ex (hunk 3 above)
Comment (mickel8) [reply lib/beamchmark.ex:63]: I don't agree :stuck_out_tongue:  I really like this syntax :heart_eyes:

File: lib/beamchmark.ex [hunk 7]
```diff
@@ -1,18 +1,135 @@
...
+  At the moment, the main interest of `#{inspect(__MODULE__)}` is scheduler utilization which tells
+  how much given scheduler was busy.
+  Scheduler is busy when:
+  * Executing process code
+  * Executing linked-in driver or NIF code
+  * Executing BIFs, or any other runtime handling
+  * Garbage collecting
+  * Handling any other memory management
+
+  Scheduler utilization is measured using Erlang's [`:scheduler`](`:scheduler`) module which uses `:erlang.statistics/1`
```
Comment (Qizot) [inline lib/beamchmark.ex:20]: Does the `[':scheduler...` thing links to erlang documentation? If so then it's really cool and strange at the same time...

File: lib/beamchmark.ex [hunk 8]
```diff
@@ -1,18 +1,135 @@
 defmodule Beamchmark do
   @moduledoc """
-  Documentation for `Beamchmark`.
+  Top level module providing `Beamchmark.run` API.
```
Comment (Qizot) [inline lib/beamchmark.ex:3]: Really nice docs :)

File: lib/beamchmark.ex (hunk 3 above)
Comment (Qizot) [reply lib/beamchmark.ex:63]: So be it  😛

File: lib/beamchmark.ex [hunk 9]
```diff
@@ -1,18 +1,135 @@
...
+    output_dir = opts[:output_dir] || @default_output_dir
+    base_dir = Path.join(output_dir, "base")
+    new_dir = Path.join(output_dir, "new")
+
+    if File.exists?(new_dir) do
+      File.rm_rf!(base_dir)
+      File.rename!(new_dir, base_dir)
+    end
+
+    Mix.shell().info("Running scenario")
```
Comment (Qizot) [inline lib/beamchmark.ex:72]: nitpick: I would add some new lines here so it does not look like a big blob, I don't mean after each line but where it makes sense.

File: lib/beamchmark.ex [hunk 10]
```diff
@@ -1,18 +1,135 @@
...
+    task = Task.async(fn -> scenario.run() end)
+    Process.sleep(opts[:delay] || 0)
```
Comment (Qizot) [inline lib/beamchmark.ex]: Can we put all the `opts` accesses at the top of the functions including their defaults?

File: lib/beamchmark.ex [hunk 11]
```diff
@@ -1,18 +1,135 @@
...
+    results = bench(opts)
+    Task.await(task, :infinity)
+    print(results, base_dir)
+    save(results, new_dir)
+    Mix.shell().info("Results successfully saved to #{inspect(new_dir)} directory")
+    :ok
+  end
+
+  defp bench(opts) do
```
Comment (Qizot) [inline lib/beamchmark.ex]: I guess this can take `duration` directly if we apply the comment above.

File: lib/beamchmark.ex [hunk 12]
```diff
@@ -1,18 +1,135 @@
...
+    Mix.shell().info("Benching")
+    Beamchmark.BEAMInfo.gather(opts[:duration] || 60)
+  end
+
+  defp save(results, new_dir) do
+    File.mkdir_p!(new_dir)
+    out = Path.join(new_dir, @results_file_name)
+    results = :erlang.term_to_binary(results)
+    File.write!(out, results)
```
Comment (Qizot) [inline lib/beamchmark.ex:95]: `File.write!` on its own returns the `:ok` atom (regarding on returning the atom below).

File: lib/beamchmark/scheduler_info.ex (hunk 6 above)
Comment (Qizot) [reply lib/beamchmark/scheduler_info.ex]: Strange...

File: lib/beamchmark/utils.ex [hunk 13]
```diff
@@ -0,0 +1,11 @@
+defmodule Beamchmark.Utils do
+  # TODO what is the return type?
```
Comment (Qizot) [inline lib/beamchmark/utils.ex]: I believe it is a `binary/String` [here](https://github.com/elixir-lang/elixir/blob/a64d42f5d3cb6c32752af9d3312897e8cd5bb7ec/lib/elixir/lib/io/ansi.ex#L4)

File: lib/beamchmark.ex (hunk 7 above)
Comment (mickel8) [reply lib/beamchmark.ex:20]: Yeah, it links to Erlang docs. I took this from Elixir source code :roll_eyes:

File: README.md [hunk 14]
```diff
@@ -1,6 +1,12 @@
...
+
+At the moment, the main interest of Beamchmark is scheduler utilization, reductions and context switches number.
+For more information please refer to API docs.
```
Comment (Qizot) [inline README.md:9]: I would probably add here or maybe in an API some information compring it to `benchee` and when you should use one and when another, but I will leave it up to you.

**Review by Qizot — APPROVED** [review]
LGTM!

**Pull request merged**